experimental/dds/tree/src/interface-prototype/Checkout.ts

```diff
@@ -0,0 +1,276 @@
+	? PlaceData
+	: TData extends RangeData
+	? RangeData
+	: TData extends TreeNodeData
+	? TreeNodeData
+	: AnchorData;
+
+//////////////////////////////////////////////////////////////////////////////////
+
+// Document and Snapshots
+
+export interface SharedTree extends Tree {
+	// createBranch(...);
+	// mergeBranch(...);
+	// deleteBranch(...);
+	// squash(firstRevision: ShortId, lastRevision: ShortId); // Revision id is changeId of preceding edit
+	// addSnapshot(revision: ShortId): SharedTreeSnapshot;
+	// getSnapshot(revision: ShortId): SharedTreeSnapshot;
+	// removeSnapshot(revision: ShortId): void;
+	// readonly firstRevision: ShortId;  // Or a constant value, e.g. 0x1?
+	// getChange(changeId: ShortId): TreeNode;
+	// nextChange(changeId: ShortId): ShortId; // Must take branching into account once we have that
+	// previousChange(changeId: ShortId): ShortId;
+	// createDiff(...);
+	// ...
+
+	// Synchronously produce a checkout where placeholders show up wherever they are in the actual data.
+	checkout(r: CommandRegistry): Checkout;
+	// Get a checkout with control over where the placeholders are.
+	checkout(r: CommandRegistry, prefetchFilter: PrefetchFilter): Promise<Checkout>;
+	// TODO: maybe option to get checkout with different API for when enforcing no placeholders at all (aka prefetching everything)
+}
+
+/**
+ * Specifies what data to prefetch.
+ * Where the filters return false, Placeholders are allowed.
+ * Returning true for all with fetch everything and produce no placeholders.
+ */
+export interface PrefetchFilter {
+	value: Filter<Definition>;
+	children: Filter<Definition>;
+	traitChunks: Filter<TraitLabel>;
+}
+
+type Filter<T> = (_: T) => boolean;
+
+interface Aborted {
+	error: Error | String;
+}
+
+/**
+ * Head of a branch. Provides APIs for:
+ *  - Viewing the version of the tree (via `Tree`). Includes locating existing Anchors in this tree.
+ *  - Notifications for changes (via `viewChange` event).
+ *  - Modifying the tree via commands which add a transaction (via `runCommand`).
+ */
+export interface Checkout extends TreeView {
+	/**
+	 * Run a command to transactionally modify the tree.
+	 *
+	 * TODO: decide if changes made during the transaction should be reflected in the output of the checkout while being applied.
+	 * Maybe sometimes, but not always?
+	 * Maybe different change event for during transactions, and/or annotate videChange with extra data about transaction?
+	 * If receiver of the event tracks the before snapshot, this pattern might work better.
+	 */
+	runCommand<TOptions extends Serializable, TAnchorSet extends AnchorSet, TResult>(
+		command: Command<TOptions, TAnchorSet, TResult>,
+		parameters: TOptions,
+		anchors: DecontextualizedAnchorSet<TAnchorSet>
+	): { editId: EditId; result: TResult } | Aborted;
+
+	// TODO: need API for async command application (runs async commands and also supports running synchronous commands with retry after loading data)
+}
```

Comment:
I think your main concepts will be easier for a TypeScript developer to follow if you hang the primary API off of a single object.  (In this case, I think that object is 'Checkout'.)

```ts
export interface Checkout {
    // read
    query(..)

    // primitive ops
    insert(..)
    detach(..)
    
    // commands
    runCommand(..)
    
    ...etc...
}
```

I think the above pivot, where anchors are passed to insert(), detach(), etc. also removes the need for "contextualized" vs. "non-contextualized" variants?

Reply:
I think you can similarly eliminate the "view" types by moving the reader APIs of checkout to a base class or interface.